zrenderer/geometry/zrbvh/rbvhtree.cpp

```diff
@@ -0,0 +1,586 @@
+            .AttachRenderbuffer( oglplus::FramebufferAttachment::Depth, depth )
+            .Complete();
+
+        proj.Set( oglplus::Mat4x3f( _transform.data(), 12 ) );
+
+        gl.Viewport( sz, sz );
+        gl.ClearColor( 0.5, 0.5f, 0.5f, 0.5f );
+        gl.Clear().ColorBuffer().DepthBuffer();
+        gl.DrawElements(
+            oglplus::PrimitiveType::Triangles,
+            _count * 3,
+            (uint32_t*)nullptr + _startIdx*3);
+
+        gl.Bound( oglplus::Framebuffer::Target::Read, fbo );
+        gl.ReadBuffer( oglplus::FramebufferColorAttachment::_0 );
+
+        _heightmap.resize( sz*sz );
+        gl.ReadPixels( 0, 0, sz, sz, oglplus::PixelDataFormat::Red,
+            oglplus::PixelDataType::Float, _heightmap.data() );
+        gl;
+    }
+
+    RBVHTrianglesPtr _triangles;
+    RBVHTriangleIndicesPtr _triangleIndices;
+    float _R, _rate;
+    Vector3f _sumAreaWeightedNormal;
+};
+
+struct RBVHNodeData
+{
+    uint32_t _data;            // 4 bytes: 3 bytes Idx, 6 bits triangle_count, 2 bits type
+    AlignedBox3h _boundingBox; // 12 bytes
+};
+
+struct RBVHTree::Impl
+{
+    Impl( ConstMeshPtr mesh, float rho, float alpha, float upperLimit )
+    {
+        MeshPtr trimesh;
+        RBVHNode nodes( mesh, trimesh, rho, alpha, upperLimit );
+
+        nodes.traverse([this]( RBVHNode* node )
+            {
+                _nodes.push_back( getNodeData( node ) );
+                if( node->_type == NT_HEIGHTMAP )
+                {
+                    _transforms.push_back( node->_transform );
+                    _heightmaps.push_back( node->_heightmap );
+                }
+            } );
+
+        _vertices = std::move( trimesh->getVertices() );
+        _indices = std::move( trimesh->getFaces() );
+
+    }
+
+    RBVHNodeData getNodeData( RBVHNode* node )
+    {
+        uint32_t data = node->_type;
+
+        switch( node->_type )
+        {
+        case NT_TRIANGLES:
+            data |= ((node->_startIdx << 8) & 0xFFFFFF00);
+            data |= ((node->_count << 3) & 0xFC);
+            break;
+        case NT_HEIGHTMAP:
+            data |= ((node->_heightIdx << 8) & 0xFFFFFF00);
+            break;
+        case NT_INTERNAL:
+            data |= ((node->_right->_nodeIdx << 8) & 0xFFFFFF00);
+            break;
+        }
+        return{
+            data,
+            Utils::halfBoundingBox( node->_bounds )};
+    }
+
+    Impl( const boost::filesystem::path& filename )
+    {
+        //TODO: fill data
+    }
+
+    void save( const boost::filesystem::path& filename )
+    {
+        boost::filesystem::ofstream
+            out( filename, boost::filesystem::ofstream::binary );
+
+        uint32_t nodeSize = (uint32_t)_nodes.size();
+        uint32_t vertSize = (uint32_t)_vertices.size();
+        uint32_t indxSize = (uint32_t)_indices.size();
+        uint32_t tranSize = (uint32_t)_transforms.size();
+
+        out.write( (char*)&nodeSize, sizeof( uint32_t ) );
```

Comment:
int8_t signed char olarak, uint8_t de unsigned char olarak ikisi de const char* e dönüştürülemiyor.

http://stackoverflow.com/questions/12381855/int8-t-and-char